src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -20952,32 +20952,1366 @@ public class C
-                // (16,9): warning CS8602: Dereference of a possibly null reference.
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field3").WithLocation(21, 9),
+                // (22,9): warning CS8602: Dereference of a possibly null reference.
                 //         field4.ToString(); // 4
-                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field4").WithLocation(16, 9)
-                );
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "field4").WithLocation(22, 9));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/56256")]
+        public void MemberNotNull_InstanceLocalFunction_TopLevel()
+        {
+            var c = CreateCompilation(["""
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+
+init();
+
+[MemberNotNull(nameof(field1))]
+void init() => throw null!;
+""", MemberNotNullAttributeDefinition]);
+
+            c.VerifyDiagnostics(
+                // (6,23): error CS0103: The name 'field1' does not exist in the current context
+                // [MemberNotNull(nameof(field1))]
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "field1").WithArguments("field1").WithLocation(6, 23));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/56256")]
+        public void MemberNotNull_InstanceLocalFunction_TopLevel_WithField()
+        {
+            var c = CreateCompilation(["""
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+
+init();
+
+[MemberNotNull(nameof(field1))]
+void init() => throw null!;
+
+public partial class Program
+{
+    public object? field1 = null;
+}
+""", MemberNotNullAttributeDefinition]);
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/56256")]
+        public void MemberNotNull_StaticLocalFunction_TopLevel()
+        {
+            var c = CreateCompilation(["""
+#nullable enable
+using System.Diagnostics.CodeAnalysis;
+
+init();
+
+[MemberNotNull(nameof(field1))]
+static void init() => throw null!;
+""", MemberNotNullAttributeDefinition]);
+
+            c.VerifyDiagnostics(
+                // (6,23): error CS0103: The name 'field1' does not exist in the current context
+                // [MemberNotNull(nameof(field1))]
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "field1").WithArguments("field1").WithLocation(6, 23));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/56256")]
+        public void MemberNotNull_EnforcedInLocalFunction_SetToMaybeNull()
+        {
+            var c = CreateNullableCompilation(["""
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    public string field1 = "";
+    public string? field2;
+    public string field3 = "";
+    public string? field4;
+
+    public void M()
+    {
+        init();
+
+        [MemberNotNull(nameof(field1), nameof(field2))]
+        void init()
+        {
+            field1.ToString(); // 1
```

Comment:
>field1.ToString(); // 1 [](http://example.com/codeflow?start=12&length=23)

Why report a warning for this case?

Reply:
Yes, that's why.
For the enforcement of `MemberNotNull` attribute within a method body, we set the initial state of the field to maybe-null when entering the body and check that it's been changed to not-null by the time you exit the method. That way we know that the method did explicitly set a not-null state. 
I used the same design for local functions. That analysis/enforcement can be performed by looking at the local function in isolation, without regard to its invocations.